sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,280 @@
+#include "otp_ctrl_regs.h"  // Generated.
+#include "rv_core_ibex_regs.h"
+
+#define IBEX_IBUS_ADDR_MATCHING_REG(i) \
+  (RV_CORE_IBEX_IBUS_ADDR_MATCHING_0_REG_OFFSET + 4 * (i))
+#define IBEX_IBUS_ADDR_EN_REG(i) \
+  (RV_CORE_IBEX_IBUS_ADDR_EN_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_DBUS_ADDR_MATCHING_REG(i) \
+  (RV_CORE_IBEX_DBUS_ADDR_MATCHING_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_ADDR_EN_REG(i) \
+  (RV_CORE_IBEX_DBUS_ADDR_EN_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_IBUS_REMAP_ADDR_REG(i) \
+  (RV_CORE_IBEX_IBUS_REMAP_ADDR_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_REMAP_ADDR_REG(i) \
+  (RV_CORE_IBEX_DBUS_REMAP_ADDR_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_IBUS_REGWEN_REG(i) \
+  (RV_CORE_IBEX_IBUS_REGWEN_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_REGWEN_REG(i) \
+  (RV_CORE_IBEX_DBUS_REGWEN_0_REG_OFFSET + 4 * (i))
+
+#define ROM_PATCH_FIELD(field, shift, length)                              \
+  enum { PATCH_REGION_##field##_SHIFT = (shift) };                         \
+  enum { PATCH_REGION_##field##_LENGTH = (length) };                       \
+  enum {                                                                   \
+    PATCH_REGION_##field##_MASK = (((~0UL) >> (32 - (length))) << (shift)) \
+  };
+
+ROM_PATCH_FIELD(LOCK_VALID, 0, 4)
+ROM_PATCH_FIELD(PROGRAM_START, 4, 4)
+ROM_PATCH_FIELD(SIZE, 8, 16)
+ROM_PATCH_FIELD(REVISION, 24, 8)
+
+#define ROM_PATCH_FIELD_VALUE(p, field)              \
+  (((((p)->header) & PATCH_REGION_##field##_MASK) >> \
+    PATCH_REGION_##field##_SHIFT))
+
+#define ROM_PATCH_MATCH_FIELD(field, shift, length)                       \
+  enum { PATCH_MATCH_##field##_SHIFT = (shift) };                         \
+  enum { PATCH_MATCH_##field##_LENGTH = (length) };                       \
+  enum {                                                                  \
+    PATCH_MATCH_##field##_MASK = (((~0UL) >> (32 - (length))) << (shift)) \
+  };
+
+ROM_PATCH_MATCH_FIELD(M_BASE, 0, 27)
+ROM_PATCH_MATCH_FIELD(P_SIZE, 27, 4)
+ROM_PATCH_MATCH_FIELD(LOCKED, 31, 1)
+
+#define ROM_PATCH_MATCH_FIELD_VALUE(m, field)       \
+  (((((m)->m_base) & PATCH_MATCH_##field##_MASK) >> \
+    PATCH_MATCH_##field##_SHIFT))
+
+enum {
+  kIbexBase = TOP_DARJEELING_RV_CORE_IBEX_CFG_BASE_ADDR,
+  kOtpBase = TOP_DARJEELING_OTP_CTRL_CORE_BASE_ADDR,
+  kSwConfig = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET,
+};
+
+enum {
+  kRomPatchBaseAddr = OTP_CTRL_PARAM_ROM_PATCH_OFFSET,
+  kRomPatchMaxAddr =
+      kRomPatchBaseAddr + (OTP_CTRL_PARAM_NUM_SW_CFG_WINDOW_WORDS * 4),
+  kRomPatchInvalidAddr = UINTPTR_MAX,
+  kRomPatchRegionHeaderSize = 4,
+  kRomPatchRegionMatchTableSize = 32 * 2 * sizeof(uint32_t),  // 32 * 2 DWORDs
+  kRomPatchRegionSignatureSize = kSigVerifyRsaNumBytes,
+  // Preamble is the patch header and the complete match table.
+  kRomPatchRegionPreambleSize =
+      kRomPatchRegionHeaderSize + kRomPatchRegionMatchTableSize,
+};
+
+#define PATCH_REGION_CODE_SIZE(p)                                       \
```

Comment:
not clear if the size is in bytes or words, I suggest to keep in words and change the name to `CODE_SIZE_WORDS`

Reply:
Made all size names specify if they're in bytes or words now, thanks for the suggestion.